### PR TITLE
make the loop:// connector built-in rather than a DSO

### PR DESCRIPTION
**Pull request by garlick**
This adds the capability for connectors to be built-in instead of requiring them all to be built as DSOs, and converts `loop://` to built-in, which allowed the duplicate implementation in libtestutil to be removed.

I started this while looking into replacing `shmem://` with a non-0MQ equivalent as a way to avoid the `zmq_ctx_term(3)` hangs when futures are leaked (#5484) .  That would need to be built-in, at least for the simple design I had in mind where named channels would be stored in a libflux-core.so global protected by mutexes, but even if that turns out to be a bad idea, this seems like a nice cleanup.

### CODE REVIEW

File: src/common/libflux/connector_loop.c [hunk 1]
```diff
@@ -87,6 +89,56 @@ static flux_msg_t *op_recv (void *impl, int flags)
     return msg;
 }
 
+static int op_getopt (void *impl, const char *option, void *val, size_t size)
+{
+    loop_ctx_t *c = impl;
+
+    if (option && streq (option, FLUX_OPT_TESTING_USERID)) {
```
Comment (grondo) [inline src/common/libflux/connector_loop.c:96]: Just curious if there's a case for this function to be called with a NULL `option`, and if not should it just have the following at the top?
```c
if (option == NULL) {
    errno = EINVAL;
    return -1;
}
```

Oh I see this just matches the usage in the local connector, but I'm still unsure of the use case for a NULL option.

**Review by grondo — APPROVED** [review]
Wow, nice cleanup! 

From the coverage reporting at least, it appears that setting testing userid/rolemask for the `loop://` connector from the environment variables `FLUX_HANDLE_USERID` and `FLUX_HANDLE_ROLEMASK` is not covered (likely wasn't before either though). Could that be dropped? If not maybe it should have a couple sanity tests? Meh, just a suggestion though.

**Comment by grondo** [comment]
Looks like codeql is failing (hopefully temporarily) due to an error fetching some ubuntu packages.

File: src/common/libflux/connector_loop.c (hunk 1 above)
Comment (garlick) [reply src/common/libflux/connector_loop.c:96]: I just copied over what was there in the `loopback_create()` implementation, but I'll have a look and see if there's a bit more cleanup we can do.

Comment (grondo) [reply src/common/libflux/connector_loop.c:96]: Seems fine to me for now if you want to defer cleanup for later. Just something that caught my eye is all..

**Comment by garlick** [comment]
> From the coverage reporting at least, it appears that setting testing userid/rolemask for the loop:// connector from the environment variables FLUX_HANDLE_USERID and FLUX_HANDLE_ROLEMASK is not covered (likely wasn't before either though). Could that be dropped? If not maybe it should have a couple sanity tests? 

The environment variables are interpreted inside `flux_open()` and translated into `flux_opt_set ()` calls of FLUX_OPT_TESTING_USERID and FLUX_OPT_TESTING_ROLEMASK.  I know there are tests that use the options direcly for the loop connector (see libflux/test/rpc_security.c for example) and sharness tests that use the environment variables with the local connector.  So I think all the code should be covered?  I'll have a closer look at the coverage report.

**Comment by grondo** [comment]
Ah, ok. I pointed it out just in case, if the code is already used and tested then it could be coverage hiccup (or more likely human error on my part) so safe to ignore it!

**Comment by garlick** [comment]
Thanks for the review!  I rebased on current master and forced a push and will set MWP.

**Pull request merged**